File: AdventOfCode_2015/Day04/rust/src/bin/part2.rs
use md5::{Digest, Md5};
fn main() {
    let secret_key = "ckczppom";
    let mut number = 1;

    loop {
        let input = format!("{}{}", secret_key, number);
        let mut hasher = Md5::new();
        hasher.update(input.as_bytes());
        let result = hasher.finalize();

        let hex_string = format!("{:x}", result);

        if hex_string.starts_with("000000") {
            println!("The lowest number is: {}", number);
            break;
        }

        number += 1;
    }
}

File: AdventOfCode_2015/Day04/java/Part2.java
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;


public class Part2 {
    public static void main(String[] args) {
        String secret_key = "ckczppom";

        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            int number = 1;
            while (true) {
                String input = secret_key + number;
                byte[] hashBytes = md.digest(input.getBytes()); // MD5 Hash
                if (startsWithFiveZeros(hashBytes)) {
                    System.out.println(number);
                    break;
                }
                number++;
            }
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    private static boolean startsWithFiveZeros(byte[] hashBytes) {
        return (hashBytes[0] == 0) && (hashBytes[1] == 0) && ((hashBytes[2]) == 0);
    }

}
